src/middlewared/middlewared/plugins/kerberos.py

```diff
@@ -0,0 +1,639 @@
+            else:
+                principal = f'{ad["bindname"]}@{ad["domainname"]}'
+                ad_kinit = await Popen(
+                    ['/usr/bin/kinit', '--renewable', '--password-file=STDIN', principal],
+                    stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE
+                )
+                output = await ad_kinit.communicate(input=ad['bindpw'].encode())
+                if ad_kinit.returncode != 0:
+                    raise CallError(f"kinit for domain [{ad['domainname']}] with password failed: {output[1].decode()}")
+        if ldap['ldap_enable'] and ldap['ldap_realm']:
+            if ldap['kerberos_principal']:
+                ad_kinit = await run(['/usr/bin/kinit', '--renewable', '-k', ldap['kerberos_principal']], check=False)
+                if ad_kinit.returncode != 0:
+                    raise CallError(f"kinit for realm {ldap['realm']} with keytab failed: {ad_kinit.stderr.decode()}")
+            else:
+                principal = f'{ldap["bindn"]}'
+                ad_kinit = await Popen(
+                    ['/usr/bin/kinit', '--renewable', '--password-file=STDIN', principal],
+                    stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE
+                )
+                output = await ad_kinit.communicate(input=ldap['bindpw'].encode())
+                if ad_kinit.returncode != 0:
+                    raise CallError(f"kinit for realm{ldap['realm']} with password failed: {output[1].decode()}")
+
+    @private
+    async def _get_cached_klist(self):
+        """
+        Try to get retrieve cached kerberos tgt info. If it hasn't been cached,
+        perform klist, parse it, put it in cache, then return it.
+        """
+        if await self.middleware.call('cache.has_key', 'KRB_TGT_INFO'):
+            return (await self.middleware.call('cache.get', 'KRB_TGT_INFO'))
+        ad = await self.middleware.call('activedirectory.config')
+        ldap = await self.middleware.call('datastore.config', 'directoryservice.ldap')
+        ad_TGT = []
+        ldap_TGT = []
+        if not ad['enable'] and not ldap['ldap_enable']:
+            return {'ad_TGT': ad_TGT, 'ldap_TGT': ldap_TGT}
+        if not ad['enable'] and not ldap['ldap_kerberos_realm']:
+            return {'ad_TGT': ad_TGT, 'ldap_TGT': ldap_TGT}
+
+        if not await self.status():
+            await self.start()
+
+        try:
+            klist = await asyncio.wait_for(
+                run(['/usr/bin/klist', '-v'], check=False, stdout=subprocess.PIPE),
+                timeout=10.0
+            )
+            if klist.returncode != 0:
+                raise CallError(f'klist failed with error: {klist.stderr.decode()}')
+        except asyncio.TimeoutError:
+            self.logger.debug('klist attempt failed after 10 seconds.')
+            await self._kdestroy
```

Comment:
I dont see a _kdestroy defined?

Reply:
It used to be. :) switched to self.stop()

Reply:
You probably also need to `return` here because you are using `klist` which may not be present if exception happens. Also asyncio timeout exception is not handled.